In [699]:
import pandas as pd
import csv
import matplotlib.pyplot as plt
import numpy as np
import seaborn as sns

In [700]:
!pwd

/Users/xhinakllogjeri/Documents/neuefische/EDA-Project/da-EDA_project


## Simple EDA 

In [701]:
df_order = pd.read_csv("data/order_data.csv", skiprows=1)
df_scan = pd.read_csv("data/order_process_data.csv")
df_cd = pd.read_csv("data/campaign_data.csv")
df_intern = pd.read_csv("data/intern_data_study.csv")

In [702]:
df_order.head()

Index        Order ID  Order Date       Ship Mode Customer ID  \
0     27  CA-2019-121755   16/1/2019    Second Class    EH-13945   
1     45  CA-2019-118255   11/3/2019     First Class    ON-18715   
2     48  CA-2019-169194   20/6/2019  Standard Class    LH-16900   
3     60  CA-2019-111682   17/6/2019     First Class    TB-21055   
4     63  CA-2018-135545  24/11/2018  Standard Class    KM-16720   

     Customer Name Origin Channel Country/Region         City       State  \
0    Eric Hoffmann          Email  United States  Los Angeles  California   
1    Odella Nelson          Sales  United States        Eagan   Minnesota   
2   Lena Hernandez          Email  United States        Dover    Delaware   
3  Ted Butterfield          Email  United States         Troy    New York   
4     Kunst Miller          Email  United States  Los Angeles  California   

   Postal Code   Region                 Category Sub-Category  \
0      90049.0     West  Special Projects Muesil  Gluten Free   
1      55122.0  Central  Special Projects Muesil  Gluten Free   
2      19901.0     East  Special Projects Muesil  Gluten Free   
3      12180.0     East  Special Projects Muesil  Gluten Free   
4      90004.0     West  Special Projects Muesil  Gluten Free   

        Product ID  Sales  Quantity  Discount   Profit  
0  TEC-AC-10003027  90.57         3       0.0  11.7741  
1  TEC-AC-10000171  45.98         2       0.0  19.7714  
2  TEC-AC-10002167  45.00         3       0.0   4.9500  
3  TEC-AC-10002167  30.00         2       0.0   3.3000  
4  TEC-AC-10004633  13.98         2       0.0   6.1512

In [703]:
df_scan.head()

Row ID        Order ID Order Date On Truck Scan Date            Ship Mode
0    3074  CA-2019-125206   3/1/2019         07/01/2019              Express
1    4919  CA-2019-160304   2/1/2019         09/01/2019  Standard Processing
2    4920  CA-2019-160304   2/1/2019         09/01/2019  Standard Processing
3    8604  US-2019-116365   3/1/2019         09/01/2019  Standard Processing
4    8605  US-2019-116365   3/1/2019         09/01/2019  Standard Processing

In [704]:
df_cd.head()

Order ID Arrival Scan Date     Customer Name
0  CA-2019-109666        03/05/2019      Kunst Miller
1  CA-2019-138933        03/05/2019       Jack Lebron
2  CA-2019-130001        03/05/2019  Heather Kirkland
3  CA-2019-113061        06/05/2019         Ed Ludwig
4  CA-2019-162138        06/05/2019       Grace Kelly

In [705]:
df_intern.head()

Order ID Ready to Ship Date Pickup Date
0  CA-2019-116540         02/09/2019  03/09/2019
1  CA-2019-116540         02/09/2019  03/09/2019
2  CA-2019-129847         04/09/2019  04/09/2019
3  CA-2019-129630         04/09/2019  04/09/2019
4  CA-2019-106278         05/09/2019  06/09/2019

In [706]:
df_order.shape

(9994, 19)

In [707]:
df_cd.shape

(333, 3)

In [708]:
df_intern.shape

(290, 3)

In [709]:
df_scan.shape

(5899, 5)

In [710]:
df_order.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9994 entries, 0 to 9993
Data columns (total 19 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Index           9994 non-null   int64  
 1   Order ID        9994 non-null   object 
 2   Order Date      9994 non-null   object 
 3   Ship Mode       9994 non-null   object 
 4   Customer ID     9994 non-null   object 
 5   Customer Name   9994 non-null   object 
 6   Origin Channel  9994 non-null   object 
 7   Country/Region  9994 non-null   object 
 8   City            9994 non-null   object 
 9   State           9994 non-null   object 
 10  Postal Code     9983 non-null   float64
 11  Region          9994 non-null   object 
 12  Category        9994 non-null   object 
 13  Sub-Category    9994 non-null   object 
 14  Product ID      9994 non-null   object 
 15  Sales           9994 non-null   float64
 16  Quantity        9994 non-null   int64  
 17  Discount        9994 non-null   f

In [711]:
df_order.columns

Index(['Index', 'Order ID', 'Order Date', 'Ship Mode', 'Customer ID',
       'Customer Name', 'Origin Channel', 'Country/Region', 'City', 'State',
       'Postal Code', 'Region', 'Category', 'Sub-Category', 'Product ID',
       'Sales', 'Quantity', 'Discount', 'Profit'],
      dtype='object')

In [712]:
df_order.columns = df_order.columns.str.replace(" ","_")
df_order.columns = df_order.columns.str.lower()
df_order.columns = df_order.columns.str.replace("/","_")
df_order.columns = df_order.columns.str.replace("-","_")

In [713]:
df_order.columns

Index(['index', 'order_id', 'order_date', 'ship_mode', 'customer_id',
       'customer_name', 'origin_channel', 'country_region', 'city', 'state',
       'postal_code', 'region', 'category', 'sub_category', 'product_id',
       'sales', 'quantity', 'discount', 'profit'],
      dtype='object')

In [714]:
df_cd.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 333 entries, 0 to 332
Data columns (total 3 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   Order ID           333 non-null    object
 1   Arrival Scan Date  333 non-null    object
 2   Customer Name      333 non-null    object
dtypes: object(3)
memory usage: 7.9+ KB


In [715]:
df_cd.columns = df_cd.columns.str.replace(" ","_")
df_cd.columns = df_cd.columns.str.lower()
df_cd.head()

order_id arrival_scan_date     customer_name
0  CA-2019-109666        03/05/2019      Kunst Miller
1  CA-2019-138933        03/05/2019       Jack Lebron
2  CA-2019-130001        03/05/2019  Heather Kirkland
3  CA-2019-113061        06/05/2019         Ed Ludwig
4  CA-2019-162138        06/05/2019       Grace Kelly

In [716]:
df_intern.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 290 entries, 0 to 289
Data columns (total 3 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   Order ID            290 non-null    object
 1   Ready to Ship Date  290 non-null    object
 2   Pickup Date         290 non-null    object
dtypes: object(3)
memory usage: 6.9+ KB


In [717]:
df_intern.columns = df_intern.columns.str.replace(" ","_")
df_intern.columns = df_intern.columns.str.lower()
df_intern.head()

order_id ready_to_ship_date pickup_date
0  CA-2019-116540         02/09/2019  03/09/2019
1  CA-2019-116540         02/09/2019  03/09/2019
2  CA-2019-129847         04/09/2019  04/09/2019
3  CA-2019-129630         04/09/2019  04/09/2019
4  CA-2019-106278         05/09/2019  06/09/2019

In [718]:
df_scan.columns = df_scan.columns.str.replace(" ","_")
df_scan.columns = df_scan.columns.str.lower()
df_scan.head()

row_id        order_id order_date on_truck_scan_date            ship_mode
0    3074  CA-2019-125206   3/1/2019         07/01/2019              Express
1    4919  CA-2019-160304   2/1/2019         09/01/2019  Standard Processing
2    4920  CA-2019-160304   2/1/2019         09/01/2019  Standard Processing
3    8604  US-2019-116365   3/1/2019         09/01/2019  Standard Processing
4    8605  US-2019-116365   3/1/2019         09/01/2019  Standard Processing

In [719]:
df_scan.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5899 entries, 0 to 5898
Data columns (total 5 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   row_id              5899 non-null   int64 
 1   order_id            5899 non-null   object
 2   order_date          5899 non-null   object
 3   on_truck_scan_date  5899 non-null   object
 4   ship_mode           5899 non-null   object
dtypes: int64(1), object(4)
memory usage: 230.6+ KB


In [720]:
df_order.describe()

index   postal_code         sales     quantity     discount  \
count  9994.000000   9983.000000   9994.000000  9994.000000  9994.000000   
mean   4997.500000  55245.233297    229.858001     3.789574     0.156203   
std    2885.163629  32038.715955    623.245101     2.225110     0.206452   
min       1.000000   1040.000000      0.444000     1.000000     0.000000   
25%    2499.250000  23223.000000     17.280000     2.000000     0.000000   
50%    4997.500000  57103.000000     54.490000     3.000000     0.200000   
75%    7495.750000  90008.000000    209.940000     5.000000     0.200000   
max    9994.000000  99301.000000  22638.480000    14.000000     0.800000   

            profit  
count  9994.000000  
mean     28.656896  
std     234.260108  
min   -6599.978000  
25%       1.728750  
50%       8.666500  
75%      29.364000  
max    8399.976000

In [721]:
df_intern.describe()

order_id ready_to_ship_date pickup_date
count              290                290         290
unique             204                 43          30
top     CA-2020-164168         09/09/2019  12/11/2020
freq                 5                 21          30

In [722]:
df_cd.describe()

order_id arrival_scan_date customer_name
count              333               333           333
unique             333               104           333
top     CA-2019-109666        03/06/2019  Kunst Miller
freq                 1                13             1

In [723]:
df_scan.describe()

row_id
count  5899.000000
mean   5022.428208
std    2878.464633
min       1.000000
25%    2485.500000
50%    5091.000000
75%    7460.000000
max    9994.000000

In [724]:
df_order.head()

index        order_id  order_date       ship_mode customer_id  \
0     27  CA-2019-121755   16/1/2019    Second Class    EH-13945   
1     45  CA-2019-118255   11/3/2019     First Class    ON-18715   
2     48  CA-2019-169194   20/6/2019  Standard Class    LH-16900   
3     60  CA-2019-111682   17/6/2019     First Class    TB-21055   
4     63  CA-2018-135545  24/11/2018  Standard Class    KM-16720   

     customer_name origin_channel country_region         city       state  \
0    Eric Hoffmann          Email  United States  Los Angeles  California   
1    Odella Nelson          Sales  United States        Eagan   Minnesota   
2   Lena Hernandez          Email  United States        Dover    Delaware   
3  Ted Butterfield          Email  United States         Troy    New York   
4     Kunst Miller          Email  United States  Los Angeles  California   

   postal_code   region                 category sub_category  \
0      90049.0     West  Special Projects Muesil  Gluten Free   
1      55122.0  Central  Special Projects Muesil  Gluten Free   
2      19901.0     East  Special Projects Muesil  Gluten Free   
3      12180.0     East  Special Projects Muesil  Gluten Free   
4      90004.0     West  Special Projects Muesil  Gluten Free   

        product_id  sales  quantity  discount   profit  
0  TEC-AC-10003027  90.57         3       0.0  11.7741  
1  TEC-AC-10000171  45.98         2       0.0  19.7714  
2  TEC-AC-10002167  45.00         3       0.0   4.9500  
3  TEC-AC-10002167  30.00         2       0.0   3.3000  
4  TEC-AC-10004633  13.98         2       0.0   6.1512

**Transform data types**  

In [725]:
df_order['order_date'] = pd.to_datetime(df_order['order_date'])

/var/folders/4t/jg4h8rnd6fxfycw0xjj3l9pw0000gn/T/ipykernel_11159/2735316145.py:1: UserWarning: Parsing dates in DD/MM/YYYY format when dayfirst=False (the default) was specified. This may lead to inconsistently parsed dates! Specify a format to ensure consistent parsing.
  df_order['order_date'] = pd.to_datetime(df_order['order_date'])


In [726]:
df_intern['ready_to_ship_date'] = pd.to_datetime(df_intern['ready_to_ship_date'])
df_intern['pickup_date'] = pd.to_datetime(df_intern['pickup_date'])

/var/folders/4t/jg4h8rnd6fxfycw0xjj3l9pw0000gn/T/ipykernel_11159/801001250.py:1: UserWarning: Parsing dates in DD/MM/YYYY format when dayfirst=False (the default) was specified. This may lead to inconsistently parsed dates! Specify a format to ensure consistent parsing.
  df_intern['ready_to_ship_date'] = pd.to_datetime(df_intern['ready_to_ship_date'])
/var/folders/4t/jg4h8rnd6fxfycw0xjj3l9pw0000gn/T/ipykernel_11159/801001250.py:2: UserWarning: Parsing dates in DD/MM/YYYY format when dayfirst=False (the default) was specified. This may lead to inconsistently parsed dates! Specify a format to ensure consistent parsing.
  df_intern['pickup_date'] = pd.to_datetime(df_intern['pickup_date'])


In [727]:
df_cd['arrival_scan_date'] = pd.to_datetime(df_cd['arrival_scan_date'])

/var/folders/4t/jg4h8rnd6fxfycw0xjj3l9pw0000gn/T/ipykernel_11159/2574422888.py:1: UserWarning: Parsing dates in DD/MM/YYYY format when dayfirst=False (the default) was specified. This may lead to inconsistently parsed dates! Specify a format to ensure consistent parsing.
  df_cd['arrival_scan_date'] = pd.to_datetime(df_cd['arrival_scan_date'])


In [730]:
df_scan['order_date'] = pd.to_datetime(df_scan['order_date'], format='%d/%m/%Y')
df_scan['on_truck_scan_date'] = pd.to_datetime(df_scan['on_truck_scan_date'],format='%d/%m/%Y')

In [732]:
df_scan.head()

row_id        order_id order_date on_truck_scan_date            ship_mode
0    3074  CA-2019-125206 2019-01-03         2019-01-07              Express
1    4919  CA-2019-160304 2019-01-02         2019-01-09  Standard Processing
2    4920  CA-2019-160304 2019-01-02         2019-01-09  Standard Processing
3    8604  US-2019-116365 2019-01-03         2019-01-09  Standard Processing
4    8605  US-2019-116365 2019-01-03         2019-01-09  Standard Processing

## Exploring and cleaning our data

In [ ]:
df_order.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9994 entries, 0 to 9993
Data columns (total 19 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   index           9994 non-null   int64         
 1   order_id        9994 non-null   object        
 2   order_date      9994 non-null   datetime64[ns]
 3   ship_mode       9994 non-null   object        
 4   customer_id     9994 non-null   object        
 5   customer_name   9994 non-null   object        
 6   origin_channel  9994 non-null   object        
 7   country_region  9994 non-null   object        
 8   city            9994 non-null   object        
 9   state           9994 non-null   object        
 10  postal_code     9983 non-null   float64       
 11  region          9994 non-null   object        
 12  category        9994 non-null   object        
 13  sub_category    9994 non-null   object        
 14  product_id      9994 non-null   object        
 15  sale

In [ ]:
df_order.drop(["postal_code", "customer_name", "city", "country_region", "region", "origin_channel", "customer_id"], axis=1, inplace=True)

In [ ]:
#checking for duplicates on order_id
df_order["order_id"].duplicated().value_counts()

False    5009
True     4985
Name: order_id, dtype: int64

We will apply the following analysis on all of our 9994 orders and not only on the unique orders.

In [ ]:
df_order.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9994 entries, 0 to 9993
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   index         9994 non-null   int64         
 1   order_id      9994 non-null   object        
 2   order_date    9994 non-null   datetime64[ns]
 3   ship_mode     9994 non-null   object        
 4   state         9994 non-null   object        
 5   category      9994 non-null   object        
 6   sub_category  9994 non-null   object        
 7   product_id    9994 non-null   object        
 8   sales         9994 non-null   float64       
 9   quantity      9994 non-null   int64         
 10  discount      9994 non-null   float64       
 11  profit        9994 non-null   float64       
dtypes: datetime64[ns](1), float64(3), int64(2), object(6)
memory usage: 937.1+ KB


**Missing values**   
Let's confirm if we have missing values again.

In [ ]:
df_order.isna().sum()

index           0
order_id        0
order_date      0
ship_mode       0
state           0
category        0
sub_category    0
product_id      0
sales           0
quantity        0
discount        0
profit          0
dtype: int64

**Build new columns/ Feature engineering**  

In [ ]:
df_scan.isna().sum()

row_id                0
order_id              0
order_date            0
on_truck_scan_date    0
ship_mode             0
dtype: int64

In [ ]:
df_cd.isna().sum()

order_id             0
arrival_scan_date    0
customer_name        0
dtype: int64

In [ ]:
df_intern.isna().sum()

order_id              0
ready_to_ship_date    0
pickup_date           0
dtype: int64

## Merging Data

In [ ]:
#df_order.merge(df_scan, how= "inner", on="order_id")

In [ ]:
df_scan.shape

(5899, 5)

In [ ]:
df_scan.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5899 entries, 0 to 5898
Data columns (total 5 columns):
 #   Column              Non-Null Count  Dtype         
---  ------              --------------  -----         
 0   row_id              5899 non-null   int64         
 1   order_id            5899 non-null   object        
 2   order_date          5899 non-null   datetime64[ns]
 3   on_truck_scan_date  5899 non-null   datetime64[ns]
 4   ship_mode           5899 non-null   object        
dtypes: datetime64[ns](2), int64(1), object(2)
memory usage: 230.6+ KB


In [733]:
df_scan["processing_time"] = (df_scan["on_truck_scan_date"]-df_scan["order_date"]).dt.days

In [734]:
df_scan

row_id        order_id order_date on_truck_scan_date  \
0       3074  CA-2019-125206 2019-01-03         2019-01-07   
1       4919  CA-2019-160304 2019-01-02         2019-01-09   
2       4920  CA-2019-160304 2019-01-02         2019-01-09   
3       8604  US-2019-116365 2019-01-03         2019-01-09   
4       8605  US-2019-116365 2019-01-03         2019-01-09   
...      ...             ...        ...                ...   
5894     908  CA-2020-143259 2020-12-30         2021-01-06   
5895     909  CA-2020-143259 2020-12-30         2021-01-06   
5896    1297  CA-2020-115427 2020-12-30         2021-01-06   
5897    1298  CA-2020-115427 2020-12-30         2021-01-06   
5898    5092  CA-2020-156720 2020-12-30         2021-01-06   

                ship_mode  processing_time  
0                 Express                4  
1     Standard Processing                7  
2     Standard Processing                7  
3     Standard Processing                6  
4     Standard Processing                6  
...                   ...              ...  
5894  Standard Processing                7  
5895  Standard Processing                7  
5896  Standard Processing                7  
5897  Standard Processing                7  
5898  Standard Processing                7  

[5899 rows x 6 columns]

In [737]:
df_scan["order_weekday"] = df_scan.order_date.dt.dayofweek

In [740]:
df_scan

row_id        order_id order_date on_truck_scan_date  \
0       3074  CA-2019-125206 2019-01-03         2019-01-07   
1       4919  CA-2019-160304 2019-01-02         2019-01-09   
2       4920  CA-2019-160304 2019-01-02         2019-01-09   
3       8604  US-2019-116365 2019-01-03         2019-01-09   
4       8605  US-2019-116365 2019-01-03         2019-01-09   
...      ...             ...        ...                ...   
5894     908  CA-2020-143259 2020-12-30         2021-01-06   
5895     909  CA-2020-143259 2020-12-30         2021-01-06   
5896    1297  CA-2020-115427 2020-12-30         2021-01-06   
5897    1298  CA-2020-115427 2020-12-30         2021-01-06   
5898    5092  CA-2020-156720 2020-12-30         2021-01-06   

                ship_mode  processing_time  order_weekday  
0                 Express                4              3  
1     Standard Processing                7              2  
2     Standard Processing                7              2  
3     Standard Processing                6              3  
4     Standard Processing                6              3  
...                   ...              ...            ...  
5894  Standard Processing                7              2  
5895  Standard Processing                7              2  
5896  Standard Processing                7              2  
5897  Standard Processing                7              2  
5898  Standard Processing                7              2  

[5899 rows x 7 columns]

In [739]:
df_scan["order_weekday"].unique() # 0 -> monday, 6-> Sunday

array([3, 2, 4, 5, 0, 1, 6])

In [742]:
df_scan["order_weekday"].value_counts() #data visualization

3    1051
2    1044
1    1031
0     970
4     885
6     593
5     325
Name: order_weekday, dtype: int64

In [735]:
# if processing > 3 check if order_date == 

In [736]:
df_scan.describe()

row_id  processing_time
count  5899.000000      5899.000000
mean   5022.428208         6.120529
std    2878.464633         2.509031
min       1.000000         0.000000
25%    2485.500000         5.000000
50%    5091.000000         7.000000
75%    7460.000000         8.000000
max    9994.000000        12.000000